routing/router_test.go

```diff
@@ -602,13 +639,157 @@ func TestSendPaymentErrorRepeatedFeeInsufficient(t *testing.T) {
 				ctx.aliases))
 	}
 }
 
+// TestSendPaymentErrorFeeInsufficientPrivateEdge tests that if we receive
+// a fee related error from a private channel that we're attempting to route
+// through, then we'll update the fees in the route hints and successfuly
+// route through the private channel in the second attempt.
+func TestSendPaymentErrorFeeInsufficientPrivateEdge(t *testing.T) {
+	t.Parallel()
+
+	const startingBlockHeight = 101
+	ctx, cleanUp, err := createTestCtxFromFile(startingBlockHeight, basicGraphFilePath)
+	defer cleanUp()
+	if err != nil {
+		t.Fatalf("unable to create router: %v", err)
+	}
+
+	// Craft a LightningPayment struct that'll send a payment from roasbeef
+	// to elst, through a private channel between son goku and elst
+	// for 1000 satoshis.
+	// This route has lower fees compared with the route through pham nuwen,
+	// as well as compared with the route through son goku -> sophon. This
+	// also holds when the private channel fee is updated to a higher value.
+	var payHash [32]byte
+	amt := lnwire.NewMSatFromSatoshis(1000)
+	privateChannelID := uint64(55555)
+	feeBaseMSat := uint32(15)
+	feeProportionalMillionths := uint32(10)
+	expiryDelta := uint16(32)
+	hopHint := HopHint{
+		NodeID:                    ctx.aliases["songoku"],
+		ChannelID:                 privateChannelID,
+		FeeBaseMSat:               feeBaseMSat,
+		FeeProportionalMillionths: feeProportionalMillionths,
+		CLTVExpiryDelta:           expiryDelta,
+	}
+	routeHints := [][]HopHint{{hopHint}}
+	payment := LightningPayment{
+		Target:      ctx.aliases["elst"],
+		Amount:      amt,
+		FeeLimit:    noFeeLimit,
+		PaymentHash: payHash,
+		RouteHints:  routeHints,
+	}
+
+	var preImage [32]byte
+	copy(preImage[:], bytes.Repeat([]byte{9}, 32))
+
+	// Prepare an error update for the private channel, with twice
+	// the original fee.
+	updatedFeeBaseMSat := feeBaseMSat * 2
+	updatedFeeProportionalMillionths := feeProportionalMillionths
+	errChanUpdate := lnwire.ChannelUpdate{
+		ShortChannelID: lnwire.NewShortChanIDFromInt(privateChannelID),
+		Timestamp:      uint32(testTime.Add(time.Minute).Unix()),
+		BaseFee:        updatedFeeBaseMSat,
+		FeeRate:        updatedFeeProportionalMillionths,
+		TimeLockDelta:  expiryDelta,
+	}
+
+	err = signErrChanUpdate(ctx.privKeys["songoku"], &errChanUpdate)
+	if err != nil {
+		t.Fatalf("Failed to sign channel update error: %v ", err)
+	}
+
+	// The error will be returned by son goku, before forwarding to the
+	// private channel, to simulate an increase in fee on the private
+	// channel that wasn't propagated back to the network.
+	sourceNode := ctx.aliases["songoku"]
+
+	// We'll now modify the SendToSwitch method to return an error on
+	// the first attempt, and check that the correct route and fees
+	// where used the second attempt.
+	errorReturned := false
+	ctx.router.cfg.SendToSwitch = func(firstHop lnwire.ShortChannelID,
+		_ *lnwire.UpdateAddHTLC, _ *sphinx.Circuit) ([32]byte, error) {
+
+		if !errorReturned {
```

Comment:
Flipping the boolean looks quite implementation-dependent. Is it possible to do it in a more robust way? For example by inspecting the contents of the `UpdateAddHTLC` and matching with the updated policy?

Reply:
What I wanted there is a simple way to send an error only the first time; that is, if its wasn't Go, I would have a static variable named say 'once' and change it to True inside the function. I think inspecting the content of UpdateAddHTLC would obfuscates this. Maybe the name errorReturned is confusing? Is there a better way in Go to achieve this 'static' variable 'do once' logic?

Reply:
Oh no, I don't mean that there is anything wrong with this code on the language level. 

This SendToSwitch is supposed to simulate the real behaviour to cover the private channel policy update case. If you flip the boolean, you rely on the fact the two htlcs will be sent out in a particular order. That's what I meant by '(router) implementation specific'. I was just thinking, if you just check the total amount in `UpdateAddHtlc` and decide based on that to reject or not, it is a little more robust.

Reply:
I don't agree here.
This is a unit test that orchestrate a scenario in which the first attempt to pay returns an error while the second one does not.
It's explicitly designed to fail only once, and not say, to fail twice and then succeed. 
Once the second attempt succeeds, the resulting route is checked, including a check that the same route is taken, and that the fee was indeed updated according to the error simulated.
I do 'rely on the fact the two htlcs will be sent out in a particular order'. Anything else will be detected as error. 
If instead I base the check on values in UpdateAddHtlc, I obfuscate the scenario I wanted to simulate (error only once), I have to avoid loops if the expected values do not appear, have to count the number of attempts to send to switch and compare later to 2, etc.

Reply:
It indeed depends on what exactly you want to assert. Ok, no change needed then.